src/config.ts

```diff
@@ -0,0 +1,246 @@
+          " instead"
+      );
+    }
+
+    if (_.has(this._src, "rules")) {
+      _.set(this._src, "database.rules", this._src.rules);
+    }
+
+    Config.MATERIALIZE_TARGETS.forEach((target) => {
+      if (_.get(this._src, target)) {
+        _.set(this.data, target, this._materialize(target));
+      }
+    });
+
+    // auto-detect functions from package.json in directory
+    if (
+      this.projectDir &&
+      !this.get("functions.source") &&
+      fsutils.fileExistsSync(this.path("functions/package.json"))
+    ) {
+      this.set("functions.source", "functions");
+    }
+  }
+
+  _hasDeepKey(obj: any, key: string) {
+    if (_.has(obj, key)) {
+      return true;
+    }
+
+    for (const k in obj) {
+      if (_.isPlainObject(obj[k]) && this._hasDeepKey(obj[k], key)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  _materialize(target: string) {
+    const val = _.get(this._src, target);
+    if (_.isString(val)) {
+      let out = this._parseFile(target, val);
+      // if e.g. rules.json has {"rules": {}} use that
+      const lastSegment = _.last(target.split("."));
+      if (_.size(out) === 1 && _.has(out, lastSegment)) {
+        out = out[lastSegment];
+      }
+      return out;
+    } else if (_.isPlainObject(val) || _.isArray(val)) {
+      return val;
+    }
+
+    throw new FirebaseError('Parse Error: "' + target + '" must be object or import path', {
+      exit: 1,
+    });
+  }
+
+  _parseFile(target: string, filePath: string) {
+    const fullPath = resolveProjectPath(this.options, filePath);
+    const ext = path.extname(filePath);
+    if (!fsutils.fileExistsSync(fullPath)) {
+      throw new FirebaseError("Parse Error: Imported file " + filePath + " does not exist", {
+        exit: 1,
+      });
+    }
+
+    switch (ext) {
+      case ".json":
+        if (target === "database") {
+          this.notes.databaseRules = "json";
+        } else if (target === "database.rules") {
+          this.notes.databaseRulesFile = filePath;
+          try {
+            return fs.readFileSync(fullPath, "utf8");
+          } catch (e) {
+            if (e.code === "ENOENT") {
+              throw new FirebaseError(`File not found: ${fullPath}`, { original: e });
+            }
+            throw e;
+          }
+        }
+        return loadCJSON(fullPath);
+      /* istanbul ignore-next */
+      case ".bolt":
+        if (target === "database") {
+          this.notes.databaseRules = "bolt";
+        }
+        return parseBoltRules(fullPath);
+      default:
+        throw new FirebaseError(
+          "Parse Error: " + filePath + " is not of a supported config file type",
+          { exit: 1 }
+        );
+    }
+  }
+
+  get src(): FirebaseConfig {
+    // TODO(samstern): We should do JSON Schema validation on this at load time
+    // and then make the _src type stronger.
+    return this._src as FirebaseConfig;
+  }
+
+  get(key: string, fallback?: any) {
```

Comment:
Totally not the right PR for this level of change, but I'd really love to get rid of all these untyped methods in the future.

Reply:
Agreed!